src/main/java/org/zalando/nakadi/domain/BatchFactory.java

```diff
@@ -1,16 +1,62 @@
 package org.zalando.nakadi.domain;
 
-import org.json.JSONArray;
+import org.json.JSONException;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class BatchFactory {
 
-    public static List<BatchItem> from(final JSONArray events) {
-        final List<BatchItem> batch = new ArrayList<>(events.length());
-        for (int i = 0; i < events.length(); i++) {
-            batch.add(new BatchItem(events.getJSONObject(i)));
+    public static List<BatchItem> from(final String events) {
+        final List<BatchItem> batch = new ArrayList<>();
+        StringBuilder sb = new StringBuilder();
```

Comment:
I was wondering if isn't there a way to build this substring with zero copy. I couldn't find, because since Java 7 String.substring creates a copy of the original char array.

I think it would be valuable to look at `JSONTokener` and try to figure out if isn't there a way to avoid extra copies of this string. This is the most executed code path in the project. I would expect that this change has an impact on garbage collection and overall performance.